# **Proyecto Waze**
**Curso 2 - Introducción a Python**


¡Bienvenido al Proyecto Waze!

Su equipo de análisis de datos de Waze aún se encuentra en las primeras etapas de su proyecto sobre abandono de usuarios. Anteriormente, se le solicitó completar una propuesta de proyecto a su supervisora, May Santner. Ha recibido la notificación de que su propuesta ha sido aprobada y que su equipo ha obtenido acceso a los datos de usuarios de Waze. Para obtener conclusiones claras, los datos de usuarios deben inspeccionarse y prepararse para el próximo proceso de análisis exploratorio de datos (EDA).

Se ha preparado un cuaderno (notebook) en Python para guiarle a través de este proyecto. Responda las preguntas y elabore un resumen ejecutivo para el equipo de datos de Waze.


# **Proyecto final del Curso 2: Inspeccionar y analizar datos**

En esta actividad, examinará los datos proporcionados y los preparará para el análisis. Esta actividad ayudará a garantizar que la información esté,

1.   Lista para responder preguntas y generar conocimientos

2.   Lista para visualizaciones

3.   Lista para futuras pruebas de hipótesis y métodos estadísticos
<br/>

**El propósito** de este proyecto es investigar y comprender los datos proporcionados.

**El objetivo** es usar un dataframe construido en Python, realizar una inspección preliminar del conjunto de datos proporcionado e informar a los miembros del equipo sobre sus hallazgos.
<br/>

*Esta actividad tiene tres partes:*

**Parte 1:** Entender la situación
* ¿Cómo puede prepararse mejor para comprender y organizar la información proporcionada?*

**Parte 2:** Entender los datos

* Crear un dataframe de pandas para aprendizaje de datos, futuros análisis exploratorios (EDA) y actividades estadísticas

* Compilar información resumida sobre los datos para informar los siguientes pasos

**Parte 3:** Entender las variables

* Utilizar los conocimientos obtenidos del examen de los datos resumidos para guiar una investigación más profunda de las variables


<br/>

Siga las instrucciones y responda las preguntas para completar la actividad. Luego, deberá completar un Resumen Ejecutivo usando las preguntas listadas en el Documento de Estrategia PACE.

Asegúrese de completar esta actividad antes de continuar. El siguiente elemento del curso le proporcionará un ejemplar completado para comparar con su propio trabajo.



# **Identificar tipos de datos y compilar información resumida**


<img src="images/Pace.png" width="100" height="100" align=left>

# **Etapas PACE**


A lo largo de estos cuadernos del proyecto, verá referencias al marco de resolución de problemas PACE. Los siguientes componentes del cuaderno están etiquetados con las etapas respectivas de PACE: Plan (Planificar), Analyze (Analizar), Construct (Construir) y Execute (Ejecutar).


<img src="images/Plan.png" width="100" height="100" align=left>

## **PACE: Plan**

Considere las preguntas en su Documento de Estrategia PACE y las que siguen para elaborar su respuesta:


### **Tarea 1. Entender la situación**

*   ¿Cómo puede prepararse mejor para comprender y organizar los datos de conductores proporcionados?*


*Comience explorando su conjunto de datos y considere revisar el Diccionario de Datos.*


**Respuesta:**

Prepárese leyendo los datos, consultando el diccionario de datos y explorando el conjunto para identificar las variables clave para el interesado.


<img src="images/Analyze.png" width="100" height="100" align=left>

## **PACE: Analyze**

Considere las preguntas en su Documento de Estrategia PACE para reflexionar sobre la etapa Analyze.


### **Tarea 2a. Importaciones y carga de datos**

















Comience importando los paquetes que necesitará para cargar y explorar el conjunto de datos. Asegúrese de usar las siguientes sentencias de importación:

*   `import pandas as pd`

*   `import numpy as np`


In [1]:
# Importar paquetes para manipulación de datos
import pandas as pd
import numpy as np


Luego, cargue el conjunto de datos en un dataframe. Crear un dataframe le ayudará a realizar manipulación de datos, análisis exploratorio (EDA) y actividades estadísticas.

**Nota:** Como se muestra en esta celda, el conjunto de datos se ha cargado automáticamente para usted. No necesita descargar el archivo .csv ni proporcionar más código para acceder al conjunto de datos y continuar con este laboratorio. Continúe con esta actividad completando las instrucciones siguientes.


In [2]:
# Cargar el conjunto de datos en un dataframe
df = pd.read_csv('waze_dataset.csv')


### **Tarea 2b. Información resumida**

Vea e inspeccione la información resumida del dataframe codificando lo siguiente:

1.   df.head(10)
2.   df.info()

*Considere las siguientes preguntas:*

1. Al revisar la salida de `df.head()`, ¿hay variables con valores faltantes?

2. Al revisar la salida de `df.info()`, ¿cuáles son los tipos de datos? ¿Cuántas filas y columnas hay?

3. ¿El conjunto de datos tiene valores faltantes?


In [3]:
df.head(10)


,ID,label,sessions,drives,total_sessions,n_days_after_onboarding,total_navigations_fav1,total_navigations_fav2,driven_km_drives,duration_minutes_drives,activity_days,driving_days,device
0,0,retained,283,226,296.748273,2276,208,0,2628.845068,1985.775061,28,19,Android
1,1,retained,133,107,326.896596,1225,19,64,13715.920550,3160.472914,13,11,iPhone
2,2,retained,114,95,135.522926,2651,0,0,3059.148818,1610.735904,14,8,Android
3,3,retained,49,40,67.589221,15,322,7,913.591123,587.196542,7,3,iPhone
4,4,retained,84,68,168.247020,1562,166,5,3950.202008,1219.555924,27,18,Android
5,5,retained,113,103,279.544437,2637,0,0,901.238699,439.101397,15,11,iPhone
6,6,retained,3,2,236.725314,360,185,18,5249.172828,726.577205,28,23,iPhone
7,7,retained,39,35,176.072845,2999,0,0,7892.052468,2466.981741,22,20,iPhone
8,8,retained,57,46,183.532018,424,0,26,2651.709764,1594.342984,25,20,Android
9,9,churned,84,68,244.802115,2997,72,0,6043.460295,2341.838528,7,3,iPhone


In [4]:
df.info()


<class 'pandas.DataFrame'>
RangeIndex: 14999 entries, 0 to 14998
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   ID                       14999 non-null  int64  
 1   label                    14299 non-null  str    
 2   sessions                 14999 non-null  int64  
 3   drives                   14999 non-null  int64  
 4   total_sessions           14999 non-null  float64
 5   n_days_after_onboarding  14999 non-null  int64  
 6   total_navigations_fav1   14999 non-null  int64  
 7   total_navigations_fav2   14999 non-null  int64  
 8   driven_km_drives         14999 non-null  float64
 9   duration_minutes_drives  14999 non-null  float64
 10  activity_days            14999 non-null  int64  
 11  driving_days             14999 non-null  int64  
 12  device                   14999 non-null  str    
dtypes: float64(3), int64(8), str(2)
memory usage: 1.7 MB


**Respuestas:**

1. Ninguna de las variables en las primeras 10 observaciones tiene valores faltantes. Tenga en cuenta que esto no implica que el conjunto de datos completo no tenga valores faltantes.

2. Las variables `label` y `device` son de tipo `object`; `total_sessions`, `driven_km_drives` y `duration_minutes_drives` son de tipo `float64`; el resto de las variables son de tipo `int64`. Hay 14,999 filas y 13 columnas.

3. El conjunto de datos tiene 700 valores faltantes en la columna `label`.


### **Tarea 2c. Valores nulos y estadística resumida**

Compare las estadísticas resumidas de las 700 filas que tienen etiquetas faltantes con las estadísticas resumidas de las filas que no tienen valores faltantes.

**Pregunta:** ¿Existe una diferencia discernible entre las dos poblaciones?


In [5]:
# Aislar filas con valores nulos
null_df = df[df['label'].isnull()]
# Mostrar estadísticas resumidas de las filas con valores nulos
null_df.describe()


,ID,sessions,drives,total_sessions,n_days_after_onboarding,total_navigations_fav1,total_navigations_fav2,driven_km_drives,duration_minutes_drives,activity_days,driving_days
count,700.000000,700.000000,700.000000,700.000000,700.000000,700.000000,700.000000,700.000000,700.000000,700.000000,700.000000
mean,7405.584286,80.837143,67.798571,198.483348,1709.295714,118.717143,30.371429,3935.967029,1795.123358,15.382857,12.125714
std,4306.900234,79.987440,65.271926,140.561715,1005.306562,156.308140,46.306984,2443.107121,1419.242246,8.772714,7.626373
min,77.000000,0.000000,0.000000,5.582648,16.000000,0.000000,0.000000,290.119811,66.588493,0.000000,0.000000
25%,3744.500000,23.000000,20.000000,94.056340,869.000000,4.000000,0.000000,2119.344818,779.009271,8.000000,6.000000
50%,7443.000000,56.000000,47.500000,177.255925,1650.500000,62.500000,10.000000,3421.156721,1414.966279,15.000000,12.000000
75%,11007.000000,112.250000,94.000000,266.058022,2508.750000,169.250000,43.000000,5166.097373,2443.955404,23.000000,18.000000
max,14993.000000,556.000000,445.000000,1076.879741,3498.000000,1096.000000,352.000000,15135.391280,9746.253023,31.000000,30.000000


In [6]:
# Aislar filas sin valores nulos
not_null_df = df[~df['label'].isnull()]
# Mostrar estadísticas resumidas de las filas sin valores nulos
not_null_df.describe()


,ID,sessions,drives,total_sessions,n_days_after_onboarding,total_navigations_fav1,total_navigations_fav2,driven_km_drives,duration_minutes_drives,activity_days,driving_days
count,14299.000000,14299.000000,14299.000000,14299.000000,14299.000000,14299.000000,14299.000000,14299.000000,14299.000000,14299.000000,14299.000000
mean,7503.573117,80.623820,67.255822,189.547409,1751.822505,121.747395,29.638296,4044.401535,1864.199794,15.544653,12.182530
std,4331.207621,80.736502,65.947295,136.189764,1008.663834,147.713428,45.350890,2504.977970,1448.005047,9.016088,7.833835
min,0.000000,0.000000,0.000000,0.220211,4.000000,0.000000,0.000000,60.441250,18.282082,0.000000,0.000000
25%,3749.500000,23.000000,20.000000,90.457733,878.500000,10.000000,0.000000,2217.319909,840.181344,8.000000,5.000000
50%,7504.000000,56.000000,48.000000,158.718571,1749.000000,71.000000,9.000000,3496.545617,1479.394387,16.000000,12.000000
75%,11257.500000,111.000000,93.000000,253.540450,2627.500000,178.000000,43.000000,5299.972162,2466.928876,23.000000,19.000000
max,14998.000000,743.000000,596.000000,1216.154633,3500.000000,1236.000000,415.000000,21183.401890,15851.727160,31.000000,30.000000


**Respuesta:**

> Comparar las estadísticas resumidas de las observaciones con etiquetas de retención faltantes con aquellas que no tienen valores faltantes no revela nada notable. Las medias y desviaciones típicas son bastante consistentes entre los dos grupos.


### **Tarea 2d. Valores nulos - conteo por dispositivo**

A continuación, verifique las dos poblaciones con respecto a la variable `device`.

**Pregunta:** ¿Cuántos usuarios de iPhone tenían valores nulos y cuántos usuarios de Android tenían valores nulos?


In [7]:
# Obtener conteo de valores nulos por dispositivo
null_df['device'].value_counts()


device
iPhone     447
Android    253
Name: count, dtype: int64

**Respuesta:**
> De las 700 filas con valores nulos, 447 eran usuarios de iPhone y 253 eran usuarios de Android.


Ahora, de las filas con valores nulos, calcule el porcentaje de cada dispositivo—Android e iPhone. Puede hacerlo directamente con la función [`value_counts()`](https://pandas.pydata.org/docs/reference/api/pandas.Series.value_counts.html).


In [8]:
# Calcular % de nulos por iPhone y por Android
null_df['device'].value_counts(normalize=True)


device
iPhone     0.638571
Android    0.361429
Name: proportion, dtype: float64

¿Cómo se compara esto con la proporción de dispositivos en el conjunto de datos completo?


In [9]:
# Calcular % de usuarios iPhone y Android en el conjunto completo
df['device'].value_counts(normalize=True)


device
iPhone     0.644843
Android    0.355157
Name: proportion, dtype: float64

El porcentaje de valores faltantes por dispositivo es consistente con su representación en el conjunto de datos en general.

No hay nada que sugiera una causa no aleatoria de los datos faltantes.


Examine los conteos y porcentajes de usuarios que abandonaron frente a los que se retuvieron. ¿Cuántos de cada grupo están representados en los datos?


In [10]:
# Calcular conteos de usuarios que abandonaron vs. retenidos
print(df['label'].value_counts())
print()
print(df['label'].value_counts(normalize=True))


label
retained    11763
churned      2536
Name: count, dtype: int64

label
retained    0.822645
churned     0.177355
Name: proportion, dtype: float64


Este conjunto de datos contiene 82% de usuarios retenidos y 18% de usuarios que abandonaron.

A continuación, compare las medianas de cada variable para usuarios que abandonaron y usuarios retenidos. La razón para calcular la mediana y no la media es que no se desea que los valores atípicos afecten indebidamente la descripción de un usuario típico. Observe, por ejemplo, que el valor máximo en la columna `driven_km_drives` es 21,183 km. ¡Eso es más de la mitad de la circunferencia de la Tierra!


In [11]:
# Calcular las medianas de todas las columnas para usuarios que abandonaron y usuarios retenidos
df.groupby('label').median(numeric_only=True)


,ID,sessions,drives,total_sessions,n_days_after_onboarding,total_navigations_fav1,total_navigations_fav2,driven_km_drives,duration_minutes_drives,activity_days,driving_days
label,,,,,,,,,,,
churned,7477.5,59.0,50.0,164.339042,1321.0,84.5,11.0,3652.655666,1607.183785,8.0,6.0
retained,7509.0,56.0,47.0,157.586756,1843.0,68.0,9.0,3464.684614,1458.046141,17.0,14.0


Esto ofrece una instantánea interesante de los dos grupos, usuarios que abandonaron vs. usuarios retenidos:

Los usuarios que abandonaron tuvieron en promedio ~3 viajes más en el último mes que los usuarios retenidos, pero los usuarios retenidos usaron la aplicación en más del doble de días que los usuarios que abandonaron en el mismo periodo.

El usuario mediano que abandonó condujo ~200 km más y 2.5 horas más durante el último mes que el usuario mediano retenido.

Parece que los usuarios que abandonaron hicieron más viajes en menos días, y sus viajes fueron más largos y de mayor duración. Quizás esto sugiera un perfil de usuario. ¡Continúe explorando!



Calcule la mediana de kilómetros por viaje en el último mes para usuarios retenidos y usuarios que abandonaron.

Comience dividiendo la columna `driven_km_drives` entre `drives`. Luego agrupe los resultados por `label` y calcule la mediana de km/viaje de cada grupo.


In [12]:
# Añadir columna `km_per_drive` al df
df['km_per_drive'] = df['driven_km_drives'] / df['drives']

# Agrupar por `label`, calcular la mediana y aislar km por viaje
median_km_per_drive = df.groupby('label').median(numeric_only=True)[['km_per_drive']]
median_km_per_drive


,km_per_drive
label,
churned,74.109416
retained,75.014702


El usuario mediano retenido condujo aproximadamente un kilómetro más por viaje que el usuario mediano que abandonó. ¿Cuántos kilómetros por día de conducción fue esto?

Para calcular esta estadística, repita los pasos anteriores usando `driving_days` en lugar de `drives`.


In [13]:
# Añadir columna `km_per_driving_day` al df
df['km_per_driving_day'] = df['driven_km_drives'] / df['driving_days']

# Agrupar por `label`, calcular la mediana y aislar km por día de conducción
median_km_per_driving_day = df.groupby('label').median(numeric_only=True)[['km_per_driving_day']]
median_km_per_driving_day


,km_per_driving_day
label,
churned,697.541999
retained,289.549333


Ahora calcule la mediana del número de viajes por día de conducción para cada grupo.


In [14]:
# Añadir columna `drives_per_driving_day` al df
df['drives_per_driving_day'] = df['drives'] / df['driving_days']

# Agrupar por `label`, calcular la mediana y aislar viajes por día de conducción
median_drives_per_driving_day = df.groupby('label').median(numeric_only=True)[['drives_per_driving_day']]
median_drives_per_driving_day


,drives_per_driving_day
label,
churned,10.0000
retained,4.0625


El usuario mediano que abandonó condujo 698 kilómetros cada día que condujo el último mes, lo que equivale a ~240% de la distancia por día de conducción de los usuarios retenidos. El usuario mediano que abandonó también tuvo un número desproporcionado de viajes por día de conducción en comparación con los usuarios retenidos.

Está claro por estas cifras que, independientemente de si un usuario abandonó o no, los usuarios representados en estos datos son conductores intensivos. Probablemente sería razonable suponer que estos datos no representan a los conductores típicos en general. Quizás los datos —y en particular la muestra de usuarios que abandonaron— contienen una alta proporción de conductores de larga distancia.

En consideración de cuánto conducen estos usuarios, valdría la pena recomendar a Waze que recopile más datos sobre estos ‘súper conductores’. Es posible que la razón por la que conducen tanto sea también la razón por la cual la aplicación Waze no satisface sus necesidades específicas, que podrían diferir de las necesidades de un conductor más típico, como un conductor que se desplaza diariamente al trabajo.


Finalmente, examine si existe un desequilibrio en cuántos usuarios abandonaron según el tipo de dispositivo.

Comience obteniendo los conteos totales de cada tipo de dispositivo para cada grupo, usuarios que abandonaron y usuarios retenidos.


In [15]:
# Para cada label, calcular el número de usuarios Android y iPhone
df.groupby(['label', 'device']).size()


label     device 
churned   Android     891
          iPhone     1645
retained  Android    4183
          iPhone     7580
dtype: int64

Ahora, dentro de cada grupo, usuarios que abandonaron y retenidos, calcule qué porcentaje era Android y qué porcentaje era iPhone.


In [16]:
# Para cada label, calcular el porcentaje de usuarios Android y iPhone
df.groupby('label')['device'].value_counts(normalize=True)


label     device 
churned   iPhone     0.648659
          Android    0.351341
retained  iPhone     0.644393
          Android    0.355607
Name: proportion, dtype: float64

La proporción de usuarios iPhone y Android es consistente entre el grupo de usuarios que abandonaron y el grupo retenido, y dichas proporciones son consistentes con la proporción encontrada en el conjunto de datos completo.


<img src="images/Construct.png" width="100" height="100" align=left>

## **PACE: Construct**

**Nota**: La etapa Construct no se aplica a este flujo de trabajo. El marco PACE puede adaptarse para ajustarse a los requisitos específicos de cualquier proyecto.




<img src="images/Execute.png" width="100" height="100" align=left>

## **PACE: Execute**

Considere las preguntas en su Documento de Estrategia PACE y las que siguen para elaborar su respuesta:


### **Tarea 3. Conclusión**

Recuerde que su supervisora, May Santer, le pidió que compartiera sus hallazgos con el equipo de datos en un resumen ejecutivo. Considere las siguientes preguntas al preparar su resumen. Piense en los puntos clave que desea compartir con el equipo y en qué información es más relevante para el proyecto de abandono de usuarios.

**Preguntas:**

1. ¿Los datos contenían valores faltantes? ¿Cuántos y qué variables se vieron afectadas? ¿Existió algún patrón en los datos faltantes?

> *El conjunto de datos tiene 700 valores faltantes en la columna `label`. No hubo un patrón obvio en los valores faltantes.*

2. ¿Cuál es una ventaja de usar la mediana en lugar de la media?

> *La media está sujeta a la influencia de valores atípicos, mientras que la mediana representa el valor central de la distribución independientemente de valores atípicos.*

3. ¿Su investigación originó preguntas adicionales que le gustaría explorar o preguntar al equipo de Waze?

> *Sí. Por ejemplo, el usuario mediano que abandonó condujo 698 kilómetros cada día que condujo el último mes, lo que es aproximadamente 240% de la distancia por día de conducción de los usuarios retenidos. Sería útil conocer cómo se recopilaron estos datos y si representan una muestra no aleatoria de usuarios.*

4. ¿Qué porcentaje de los usuarios en el conjunto de datos eran usuarios Android y qué porcentaje eran usuarios iPhone?

> *Los usuarios Android representaron aproximadamente 36% de la muestra, mientras que los usuarios iPhone representaron alrededor de 64%.*

5. ¿Cuáles fueron algunas características distintivas de los usuarios que abandonaron respecto a los usuarios retenidos?

> *En general, los usuarios que abandonaron condujeron más lejos y por más tiempo en menos días que los usuarios retenidos. También usaron la aplicación aproximadamente la mitad de las veces que los usuarios retenidos durante el mismo periodo.*

6. ¿Hubo una diferencia apreciable en la tasa de abandono entre usuarios iPhone y Android?

> *No. La tasa de abandono para iPhone y Android estuvo dentro de un punto porcentual entre sí. No hay indicios de que el abandono esté correlacionado con el dispositivo.*


**¡Felicitaciones!** Ha completado este laboratorio. Sin embargo, es posible que no vea una marca de verificación verde junto a este elemento en la plataforma de Coursera. Continúe su progreso independientemente de la marca. Haga clic en el icono "guardar" en la parte superior de este cuaderno para asegurarse de que su trabajo se haya registrado.
